Support load state dict in TorchNano

MeouSker77 opened the pull request:
## Description

<!-- For small changes (<=3 files and <=50 lines of codes in the source folder), -->
<!-- you may remove Sections 1-4 below and just provide a simple description here -->

### 1. Why the change?

<!-- Provide the related github issue link if available -->

### 2. User API changes

<!-- Describe API changes (i.e., how users will use the feature) if any; -->
<!-- alternatively, provide a link to the github issue for the design -->

### 3. Summary of the change 

<!-- Provide the design for the implementation; -->
<!-- alternatively, provide a link to the github issue for the design -->

### 4. How to test?
- [ ] N/A
- [ ] Unit test
- [ ] Application test
- [ ] Document test
- [ ] ...

### 5. New dependencies

<!-- If no new dependency is introduced, remove this section -->

- [ ] New Python dependencies
       - Dependency1 
       - Dependency2
       - ...
- [ ] New Java/Scala dependencies and their license
       - Dependency1 and license1
       - Dependency2 and license2
       - ...

jason-dai commented on python/nano/src/bigdl/nano/pytorch/torch_nano.py:
```diff
@@ -180,12 +243,21 @@ def setup(self, model: nn.Module,    # type: ignore[override]
+        else:
+            model, optimizers = self._setup(model, optimizers, move_to_device=move_to_device)
+            model_wrapper = _TorchNanoModuleWrapper(model, self, model,  # type: ignore
+                                                    optimizers, move_to_device)
```
1) Why not handle this in `_setup` instead?
2) I think you can extend `_TorchNanoModule` instead of introducing a new class `_TorchNanoModuleWrapper`? `_TorchNanoModule` is already a wrapper class and also deals with `__getattr__`.

MeouSker77 replied:
if we extend `_TorchNanoModule` instead of using `_TorchNanoModuleWrapper`, then the `load_state_dict` method will be:
```python
class TorchNanoModule(_LiteModule):
    def load_state_dict(self, state_dict, xxx):
        backup_model.load_state_dict(state_dict)
        new_model = xxx._setup(backup_model)
        self = new_model
```

however, `self = new_model` is invalid in python, we cann't replace `self` in the `load_state_dict` method

jason-dai replied:
> if we extend `_TorchNanoModule` instead of using `_TorchNanoModuleWrapper`, then the `load_state_dict` method will be:
> 
> ```python
> class TorchNanoModule(_LiteModule):
>     def load_state_dict(self, state_dict, ...):
>         backup_model.load_state_dict(state_dict)
>         new_model = xxx._setup(backup_model)
>         self = new_model
> ```
> 
> however, `self = new_model` is invalid in python, we can't replace `self` in the `load_state_dict` method

We can have `_setup`  to return the normal model (possibly optimized by ipex), and wrap it with `_TorchNanoModule` in `setup`; then we can have something like:

```python
class TorchNano(LightningLite):
  def setup(...)
        ...
        model, optimizers = self._setup(model, optimizers, ...)
        model = _TorchNanoModule(model, self._precision_plugin)
        ...

class TorchNanoModule(_LiteModule):
  def load_state_dict(self, state_dict, ...):
        backup.load_state_dict(state_dict)
        module = xxx._setup(backup_model)
        ...
```

jason-dai commented:
> ### 2. User API changes
> user can wirte the following code now:
> 
> ```python
> class MyNano(TorchNano):
>     def train(self):
>         ...
>         model, optimizer = self.setup(model, optimizer)
>         ...
>         model_state_dict = model.state_dict()
>         optimizer_state_dict = optimizer.state_dict()
>         ...
>         # no more training
>         model.load_state_dict(model_state_dict)
>         # still want to continue training
>         optimizer = model.load_state_dict(model_state_dict, optimizer_state_dict=optimizer_state_dict)
>         ...
> ```
> 
> * note: if user want to continue training after calling `load_state_dict`, he must pass the optimizer's state dict together with the model's state dict, and use the returned optimizer for the following training (the returned optimizer is same as the optimizer passed to `self.setup` method)

I don't think this API aligns with standard PyTorch interfaces for saving & loading the model (https://pytorch.org/tutorials/beginner/saving_loading_models.html)

jason-dai commented on python/nano/src/bigdl/nano/pytorch/torch_nano.py:
```diff
@@ -34,6 +36,63 @@
+            for optimizer, state in zip(self.backup_optimizers, optimizer_state_dict_list):
+                optimizer.load_state_dict(state)
+
+        if self.torch_nano.use_ipex and not TORCH_VERSION_LESS_1_10:
```
use a flag such as `has_backup`

MeouSker77 commented:
> I don't think this API aligns with standard PyTorch interfaces for saving & loading the model (https://pytorch.org/tutorials/beginner/saving_loading_models.html)

yes, it doesn't align with standard PyTorch, but we may have no other way.

If we replaced a model, then we have to replace the optimizer too. Standard Pytorch model loads the state dict in place, so it doesn't need a new optimizer. But we replaced the old model, and the old optimizer is stilled bound to the old model, so we need a new optimizer, which is bound to the new model.

jason-dai commented:
> > I don't think this API aligns with standard PyTorch interfaces for saving & loading the model (https://pytorch.org/tutorials/beginner/saving_loading_models.html)
> 
> yes, it doesn't align with standard PyTorch, but we may have no other way.
> 
> If we replaced a model, then we have to replace the optimizer too. Standard Pytorch model loads the state dict in place, so it doesn't need a new optimizer. But we replaced the old model, and the old optimizer is stilled bound to the old model, so we need a new optimizer, which is bound to the new model.



> > I don't think this API aligns with standard PyTorch interfaces for saving & loading the model (https://pytorch.org/tutorials/beginner/saving_loading_models.html)
> 
> yes, it doesn't align with standard PyTorch, but we may have no other way.
> 
> If we replaced a model, then we have to replace the optimizer too. Standard Pytorch model loads the state dict in place, so it doesn't need a new optimizer. But we replaced the old model, and the old optimizer is stilled bound to the old model, so we need a new optimizer, which is bound to the new model.

What type of code pattern do we want to support? Maybe we should save the non-ipex version of the model, and do not allow ipex-optimized model to load; and we will always load non-ipex version of the model, and need to call `setup` again for retraining?

```python
torch.save({
            'model_state_dict': model.state_dict(),
            'optimizer_state_dict': optimizer.state_dict(),
             ...
            }, PATH)
```
```python
model = TheModelClass(...)
optimizer = TheOptimizerClass(...)

checkpoint = torch.load(PATH)
model.load_state_dict(checkpoint['model_state_dict'])
optimizer.load_state_dict(checkpoint['optimizer_state_dict'])
...

model, optimizer = self.setup(model, optimizer, ...)
```

TheaperDeng reviewed (approved):
lgtm, we definitly need a user guide for this "trick"/"workaround" about state dict

jason-dai commented:
> lgtm, we definitly need a user guide for this "trick"/"workaround" about state dict

Need to add basic examples/how-to on model save & load (for both training and inference).
